ompi/mca/coll/han/coll_han_alltoallv.c

```diff
@@ -0,0 +1,878 @@
+                        iov.iov_len = buf_remain;
+                        iov_count = 1;
+                        opal_convertor_unpack(&recv_convertor, &iov, &iov_count, &nbytes_pack);
+
+                        buf_remain -= nbytes_pack;
+                        buf_converted += nbytes_pack;
+                        recv_convertor_bytes_remaining -= nbytes_pack;
+                    }
+                }
+            }
+
+            size_t bytes_to_post = MIN(buf_len, recv_post_remaining_bytes);
+            if (bytes_to_post > 0) {
+                /* post a new recv */
+                MCA_PML_CALL(irecv
+                    (req_buf, bytes_to_post, MPI_PACKED, jrank_recvfrom,
+                    MCA_COLL_BASE_TAG_ALLTOALLV,
+                    comm, &requests[jreq]));
+
+                /* update posted_recv_bytes */
+                recv_post_remaining_bytes -= bytes_to_post;
+            } else {
+                requests[jreq] = MPI_REQUEST_NULL;
+            }
+        }
+    }
+
+    OBJ_DESTRUCT(&send_convertor);
+    OBJ_DESTRUCT(&recv_convertor);
+
+    for (int jbuf=0; jbuf<nbufs; jbuf++) {
+        opal_free_list_return(&mca_coll_han_component.pack_buffers, buf_items[jbuf]);
+    }
+    return 0;
+}
+
+static int decide_to_use_smsc_alg(
+    int *use_smsc,
+    const void *sbuf,
+    ompi_count_array_t scounts,
+    ompi_disp_array_t sdispls,
+    struct ompi_datatype_t *sdtype,
+    void* rbuf,
+    ompi_count_array_t rcounts,
+    ompi_disp_array_t rdispls,
+    struct ompi_datatype_t *rdtype,
+    struct ompi_communicator_t *comm)
+{
+    opal_convertor_t convertor;
+    long long reduce_buf_input[3];
+    long long reduce_buf_output[3];
+    int bufs_on_device = 0;
+    size_t avg_send_bytes = 0;
+    size_t packed_size_bytes = 0;
+    int need_bufs = 0;
+    int rc;
+    int comm_size = ompi_comm_size(comm);
+    int comm_rank = ompi_comm_rank(comm);
+
+    /*
+    Perform an allgather on all ranks to decide if this algorithm is worth
+    using. There are three important things:
+
+     1. Device buffers.  XPMEM doesn't support GPU buffers, so we cannot proceed
+        with this algorithm.
+     2. Send size per rank.  This algorithm can pack small messages together,
+        but this typically isn't helpful for large messages, and XPMEM-mapped
+        memory cannot be registered with ibv_reg_mr.
+     3. Contiguous buffers.  The exception to #2 above is if we can't post our
+        sends/recvs in one large block to the NIC.  For these non-contiguous
+        datatypes, our packing algorithm is better because (a) we re-use our
+        buffers from a free-list which can remain registered to the NIC, and (b)
+        we pipeline multiple sends at once (c) we can pack more data into each
+        send than the network layer because we have mmap'ed all local ranks and
+        understand their datatypes.
+
+    The allgather adds about 50 usec on 256 ranks across 4 nodes, but for large
```

Comment:
This comment talk about an allgather but the code uses an allreduce. What did I miss ?

Reply:
My mistake: should say allreduce.  Addressed.